microsetta_admin/metadata_util.py

```diff
@@ -1,141 +1,258 @@
-                        if not answer_set[a]:
-                            error_report.append({
-                                "barcode": barcode,
-                                "error": "Invalid response to question " +
-                                         str(q_id) + "/" + str(col_name)
-                            })
-                else:
-                    if col_name in transformed_response[barcode]:
-                        error_report.append({
-                            "barcode": barcode,
-                            "error": "Duplicate response to question " +
-                                     str(q_id) + "/" + str(col_name)
-                        })
-                    else:
-                        transformed_response[barcode][col_name] = answer
-
-    # print("TEMPLATES")
-    # print(json.dumps(cached_templates, indent=2))
-    # print("RESPONSES")
-    # print(json.dumps(cached_response_json, indent=2))
-    # print("TEMPLATE POSSIBLE MULTISELECT OPTIONS")
-    # print(json.dumps(multiselect_map, indent=2))
-    # print("TRANSFORMED")
-    # print(json.dumps(transformed_response, indent=2))
-    # print("ERRORS")
-    # print(json.dumps(error_report, indent=2))
-
-    return transformed_response, error_report
-
-
-def _build_possible_values_per_question(templates):
-    final_map = {}
-
-    def visit_field(map, field):
-        if field["type"] == "checklist":
-            final_map[field["id"]] = field["values"]
-
-    for template_id in templates:
-        template = templates[template_id]
-        if template["survey_template_text"]["fields"]:
-            for field in template["survey_template_text"]["fields"]:
-                visit_field(final_map, field)
-        for group in template["survey_template_text"]["groups"]:
-            for field in group["fields"]:
-                visit_field(final_map, field)
-
-    return final_map
+
+        fetched.append(barcode_metadata)
+
+    if len(fetched) == 0:
+        error_report.append({"error": "No metadata was obtained"})
+        df = pd.DataFrame()
+    else:
+        df = _to_pandas_dataframe(fetched)
+
+    return df, error_report
+
+
+def _to_pandas_dataframe(metadatas):
+    """Convert the raw barcode metadata into a DataFrame
+
+    Parameters
+    ----------
+    metadatas : list of dict
+        The raw metadata obtained from the private API
+
+    Returns
+    -------
+    pd.DataFrame
+        The fully constructed sample metadata
+    """
+    transformed = []
+    multiselects = set()
+
+    for metadata in metadatas:
+        as_series, observed_multiselect = _to_pandas_series(metadata)
+        transformed.append(as_series)
+        multiselects.update(observed_multiselect)
+
+    df = pd.DataFrame(transformed)
+    df.index.name = 'sample_name'
+
+    # update reponse where someone did not check a multi-select box
+    for multiselect in multiselects:
+        nulls = df[multiselect].isnull()
+        df.loc[nulls, multiselect] = 'false'
```

Comment:
I believe that this will fill in the string 'false' for all options of a multiselect which were checked in at least one survey from the barcodes chosen.  However, if all barcodes involved did not check this option, that column will be stripped from the output.  This means our csvs would all have different sets of columns no?

Reply:
I think that is accurate. However I'm uncertain if it matters. It would mean that on concatenation, there would be nulls. But actually, a false (or null for that matter) could be mean either unspecified or a true negative. However, I think being consistent here is important... good spot.

Reply:
I'm going to change this to "checked" and "unchecked" which better reflects what the values are